### ISSUE TITLE
What are the CF-functionalities we need?

### ISSUE THREAD

**Issue by ocefpaf** [ioos/APIRUS#11]
@rsignell-usgs mentioned in #10 

> The CF model itself actual consists of functionality that can be separated: unit conversion, vertical coordinate calculation, standard_name manipulation, handling of different common data model featureTypes (Grid, Point, TimeSeries, TimeSeriesProfile, Profile, Trajectory, TrajectoryProfile). Grid handles only data which is colocated with coordinate values.

I would like to start a discussion on how are we on each of these topics.  Maybe we should open separate issue for each.  My take is:
- **unit conversion:** `cf_units` is out there and works.  It has all the UDUNITS idiosyncrasies but it is the standard adopt in CF and many other conventions. 
- **vertical coordinate calculation:** we have this functionality in iris and parts of it scattered among several ROMS modules out there.  I hope that [odvc](https://github.com/pyoceans/odvc) can tie all this together.
- **standard_name manipulation:** what do we need here beyond compliance-check?  Iris ships the standard_name xml and uses it to create the cube.  I am unaware of any other package that uses it.
- **handling of different common data model:** that is interesting.  If I extract a time-series from a model should that become a `TimeSeries` `featureType` automatically?  What would be the functionalities expected here?
- **grid:** this goes beyond CF and the discussion has already started involving pyugrid and pysgrid.

**Comment by kwilcox** [comment]
- **unit conversion:** `cf_units`\- done
- **vertical coordinate calculation:** `odvc` - will be done
- **standard_name manipulation:** I don't think we need this... I don't want to be checking CF compliance... we can use compliance-checker for that: https://github.com/ioos/compliance-checker/.
- **handling of different common data model:** - I think we will need an implementation of the CF DSGs in python, and it should be implemented in the same way pyugrid and pysgrid will be so the common API is consistent. I would love it if extracting a timeseries from a model produced a TimeSeries object I could then export to NetCDF (or just analyze locally in memory).

**Comment by ocefpaf** [comment]
I guess a new issue for **handling of different common data model** is in order then.

**Comment by lukecampbell** [comment]
in addition to the vertical coordinate caluclations we need to be able to support mapping to and from depth and sigmas which unfortunately requires a bathymetry database. In some cases some model outputs include the bathymetry used, but not most.

**Comment by rsignell-usgs** [comment]
@lukecampbell, if a model is using coordinates with sigma, it is supposed to provide the depth, or else it's not CF-compliant.   I think we've agreed the API should allow the user to fix non-CF datasets by having a callback or similar functionality.   So you could define a function that CF-izes your dataset (or variable) by providing the bathymetry if it was not provided.    And if you wanted to estimate the model depths by reading from a bathymetry database you could do so.

**Comment by lukecampbell** [comment]
Correct me if I'm mistaken here but currently unstructured grids are inherently non-CF compliant. At least 1.6. I'm pretty sure FVCOM uses sigma, do you know if it includes depth in the model run outputs?

**Comment by ocefpaf** [comment]
> Unstructured grids are inherently non-CF compliant.

Not at all!  One convention does convention does not exclude the other.  In fact they are additive.  The dataset below is `Conventions: CF-1.0` and `ugrid 0.9`.  (Even though the metadata does not say anything about the latter.)

> FVCOM uses sigma, do you know if it includes depth in the model run outputs?

Look for the variable `h`:

http://www.smast.umassd.edu:8080/thredds/dodsC/FVCOM/NECOFS/Forecasts/NECOFS_GOM3_FORECAST.nc.html

**Comment by lukecampbell** [comment]
That's good to know. We had a similar project to this here and I remember we had some troubles dealing with models that used sigma because we couldn't get the actual depth. This makes me wonder what model we were using that didn't include depth.

**Comment by ocefpaf** [comment]
Yoda would say:

```
   Valid your mode is not.
   Vertical coordinates without depth you cannot compute.
   Use the metadata Luke!
```

**Comment by lukecampbell** [comment]
Oh, I know. My biggest issue is the general lack of metadata.

**Comment by ocefpaf** [comment]
``` python
"""
       Valid your mode is not.
       Vertical coordinates without depth you cannot compute.
       Use the metadata Luke!
""".replace("Use the metadata Luke!", "Scream at the data provider Luke!")
```

**Comment by lukecampbell** [comment]
That's at least half of my current job is working with providers to improve their metadata. We always have the same arguments though. "Why can't you just do it?" -- Provider.  "Because I legitimately have no idea what's needed here." -- Me.

**Comment by emiliom** [comment]
I'm intrigued that @kwilcox (or anyone else) didn't mention [Paegan](https://github.com/asascience-open/paegan) here, or in #10 or #12. Admittedly, Paegan seems to be dormant (last activity was 1.5 years ago), but it did bill itself as "The Python CDM for Met/Ocean data", and it's still a/the backbone behind [pyoos](https://github.com/ioos/pyoos/). At least for TimeSeries, I imagine there's a lot to be had or extracted-from there.

Nifty discussions you've got going here and in #10! Coincidentally I learned today about Unidata's "pyCDM" project (only as a brief reference in a [recent Unidata meeting report](http://www.unidata.ucar.edu/committees/usercom/2015Sep/statusreports/status_reports_sept2015.pdf), but couldn't find any actual repo or some such tangible, public evidence for it.

**Comment by ocefpaf** [comment]
pyCDM sounds interesting.  I found [this](http://www.unidata.ucar.edu/committees/usercom/2014Sep/statusreports/python.html) and the last bullet point mention that some code exist.  Hopefully that will be on GiHub at some point.

**Comment by rsignell-usgs** [comment]
I'm betting @dopplershift or @lesserwhirls would be willing to fill us in on pyCDM thoughts and plans...

**Comment by dopplershift** [comment]
Well, not much to fill in. We'd like to get a lot of the useful functionality of the CDM into Python. Unfortunately, that's a big task with not a lot of resources for us to put into it. Us Python folks here have kept focused on much lower-hanging fruit.

Now if we have a critical mass of the community interested here, we'd of course be very interested in helping out....
